subprojects/android-test/ui-testing-core/src/main/kotlin/com/avito/android/test/matcher/BitmapMatcher.kt

```diff
@@ -17,17 +18,26 @@ class BitmapMatcher(private val bitmap: Bitmap) :
         description.appendText("with bitmap: ")
     }
 
-    override fun matchesSafely(layout: ImageView): Boolean {
-        val currentBitmap = (layout.drawable as? BitmapDrawable)?.bitmap
-        val shouldBe = bitmap.getTopLeftPixels()
-        val actual = currentBitmap?.getTopLeftPixels()
-            ?: throw AssertionFailedError("This Drawable is not Bitmap or not exist")
-
-        val i = shouldBe.iterator()
-        val j = actual.iterator()
-        while (i.hasNext() && j.hasNext()) {
-            if (!hasSameColor(i.next(), j.next())) {
-                throw AssertionFailedError("Bitmaps has different colors!")
+    override fun matchesSafely(view: ImageView): Boolean {
+        val actualBitmap: Bitmap = when (val drawable = view.drawable) {
+            is BitmapDrawable -> drawable.bitmap
+            else -> drawable.toBitmap()
+        }
+        if (actualBitmap.height != bitmap.height
+            || actualBitmap.width != bitmap.width
+        ) {
+            throw AssertionFailedError(
+                "Bitmaps has different sizes: " +
+                    "actual [${actualBitmap.height}x${actualBitmap.width}], compared [${bitmap.height}x${bitmap.width}]"
+            )
+        }
+        val actualPixels = actualBitmap.getPixelsSnapshot()
+        val shouldBePixels = bitmap.getPixelsSnapshot()
+
+        shouldBePixels.forEachIndexed { index, shouldBePixel ->
+            val actualPixel = actualPixels[index]
+            if (!hasSameColor(shouldBePixel, actualPixel)) {
+                throw AssertionFailedError("Bitmaps are different")
```

Comment:
Maybe it is possible to calculate (x, y) position of the first different pixel so it will be much easier to a developer to find the difference between images in any graphical editor.

Reply:
Agree, will add information about pixels

Reply:
Fixed